patternfly/patternfly-elements
# feat(jump-links)!: reimplement in line with PFv4

## What I did
- added `@patternfly/pfe-core/controllers/scroll-spy-controller.js`
- reimplemented pfe-jump-links with above, and with styles and APIs in line with PFv4 designs

## Testing Instructions
Run the demo locally, and check the [Deploy Preview](https://deploy-preview-2283--patternfly-elements.netlify.app/components/jump-links/demo/)
- is it accessible? 
- how quickly can you break the scroll spy? under what conditions? give clear repros, please!
- compare the styles and behaviour to [PFv4](https://www.patternfly.org/v4/components/jump-links/) and to the [Scroll spy demo](https://www.patternfly.org/v4/components/jump-links/react-demos/scrollspy-with-subsections/) ([Deploy Preview](https://deploy-preview-2283--patternfly-elements.netlify.app/components/jump-links/demo/scrollspy-with-subsections/))

## Notes to reviewers
### Public APIs/HTML
Keep in mind that the primary goal here is to *replicate the pfv4* dx, not to provide a strictly html-first element. Nonetheless, the controller tries not to presume anything about the available DOM, so that it could be useful in implementing a light-dom-heavy version of jump-links in the near future (i.e. rh-jump-links or rh-table-of-contents, whathaveyou).

Try to think ahead to what you'd need to implement an HTML-first / noscript-y version of this, and think if the controller as-is would be enough.

### Assumptions made
There are, as far as i can tell, two places where the controller makes assumptions about it's host's light DOM, both regarding this concept of "link children". Link children are the elements in the controller host's light-dom which link to the scrollable content headings. when the content scrolls up and down the page, the link child which represents the current heading should become "active". 

1. we assume that link children have an `href` attribute which contains a hash reference to its heading (i.e. `<a href="#foo">` or `<pfe-jump-links-item href="#bar">`
2. we assume that it's sufficient to toggle some attribute on the link child when its heading scrolls into view.

If you find these assumptions to be too restrictive, please speak up.
